### PR TITLE
Add constant fold to onnx dialect

### PR DESCRIPTION
**Pull request by chentong319**
This PR is the new version of #1799 on the current code base to avoid handling the merging conflicts.
This PR enables the Fold support in onnx Dialect. 
1. Dialect materializer support
2. Add fold function to SqueezeOp as an example
3. Update the lit test. A new test case is added. For SqueezeOp, the canonicalization can do what constprop does previously. The existing test is updated because the folding in canonicalization moves all the constants to the beginning of the function, and inconsistent constantOp in previous test cases cannot work anymore.

Details:
1. Fold is called automatically by canonicalization. We can move most of the work in onnx-mlir ConstProp pass, except for restructuring tree for commutative Ops, to fold function. The benefit is to use `OpBuilder::createOrFold` when Ops are generated in transformation.
2. There is some overlapping between our constant propagation and shape inference: we assume that the shape inference is done. constant propagation may use the type of Op output to generate  output constant. This is not true for canonicalization with fold. In canonicalization of ConvOp, the generated SqueezeV11Op has unranked tensor output.
3. Fold function also requires that the attribute of a ConstantOp is consistent with its output.

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp [hunk 1]
```diff
@@ -183,3 +183,39 @@ namespace onnx_mlir {
 template struct ONNXCommonSqueezeOpShapeHelper<ONNXSqueezeOp>;
 template struct ONNXCommonSqueezeOpShapeHelper<ONNXSqueezeV11Op>;
 } // namespace onnx_mlir
+
+//===----------------------------------------------------------------------===//
+// Folder
+//===----------------------------------------------------------------------===//
+OpFoldResult ONNXSqueezeOp::fold(FoldAdaptor adaptor) {
+  OnnxElementsAttrBuilder elementsBuilder(getContext());
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp:191]: better to move it down just before it's used, so it's only created when used

File: src/Dialect/ONNX/ONNX.td [hunk 2]
```diff
@@ -41,6 +41,8 @@ def ONNX_Dialect : Dialect {
   let useDefaultAttributePrinterParser = 0;
   let useFoldAPI = kEmitFoldAdaptorFolder;
   let dependentDialects = ["func::FuncDialect"];
+  let useFoldAPI = kEmitFoldAdaptorFolder;
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNX.td:44]: duplicated, already appears two lines above

File: src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp [hunk 3]
```diff
@@ -183,3 +183,39 @@ namespace onnx_mlir {
...
+  if (!adaptor.getData() || !adaptor.getAxes()) {
+    // Use original Op if Data or Axes is not constant
+    return nullptr;
+  }
+
+  // ToFix: copied from constprop.cpp.
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp:197]: this is the way to do it and there's nothing to fix, right? i.e. we shouldn't have "ToFix" in the comment

if the result shape is not static then we just cannot fold, because the constant folding breaks down if a constant op has non-static shape, as it must derive its type from the attribute, which always has static type

File: src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp [hunk 4]
```diff
@@ -183,3 +183,39 @@ namespace onnx_mlir {
...
+  // Constant propagation pass always follows shape inference,
+  // and assumes the shape is inferenced.
+  // But the canonicalization of ConvOp may generate SqueezeOp with
+  // unranked tensor output. Type checking is needed here.
+  auto ty = getOperation()->getResults()[0].getType().dyn_cast<ShapedType>();
+  if (!(ty && ty.hasStaticShape()))
+    return nullptr;
+
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp:205]: simpler: `if (!hasStaticShape(getSqueezed().getType()) return nullptr;`

File: src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp (hunk 3 above)
Comment (chentong319) [reply src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp:197]: I am thinking of calling Shape inference here to fix it.

Comment (sorenlassen) [reply src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp:197]: the issue is that if we change the operation's result type then things go hay-wire, especially if it's the last op in a function body then MLIR will complain that the return result (operand of the return op) doesn't match the function's return type, so we will need to propagate the type change to some of the users of the return value, which might have knock-on effects wherever we canonicalize ops

maybe we should do that: change the canonicalize pass etc to do the necessary shape inference and shape propagation to support constant folding, but I can't predict how many obstacles we'll encounter if we try that

Comment (chentong319) [reply src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp]: I tried to add the shape inference. It works for my test case. I will check the last op issue you mentioned.
Now `onnx-mlir-opt --canonicalize` of the following model (in onnx_canonicalization.mlir)
```
func.func @test_fuse_add_conv(%arg0 : tensor<1x1x28x28xf32>, %arg1 : tensor<8x1x5x5xf32>) -> tensor<1x8x28x28xf32> {
    %cst = "onnx.NoValue"() {value} : () -> none
    %0 = "onnx.Conv"(%arg0, %arg1, %cst) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : si64, kernel_shape = [5, 5], onnx_node_name = "Convolution28", strides = [1, 1]} : (tensor<1x1x28x28xf32>, tensor<8x1x5x5xf32>, none) -> tensor<1x8x28x28xf32>
    %1 = onnx.Constant dense<[[[-0.161539719]], [[-0.433835655]], [[0.091641359]], [[-0.0168522168]], [[-0.0650264397]], [[-0.131737873]], [[0.0204175506]], [[-0.121110231]]]> : tensor<8x1x1xf32>
    %2 = "onnx.Add"(%0, %1) : (tensor<1x8x28x28xf32>, tensor<8x1x1xf32>) -> tensor<1x8x28x28xf32> 
    return %2 : tensor<1x8x28x28xf32> 
}
```
Will generate the result 
```
module {
  func.func @test_fuse_add_conv(%arg0: tensor<1x1x28x28xf32>, %arg1: tensor<8x1x5x5xf32>) -> tensor<1x8x28x28xf32> {
    %0 = onnx.Constant dense<[-0.161539719, -0.433835655, 0.091641359, -0.0168522168, -0.0650264397, -0.131737873, 0.0204175506, -0.121110231]> : tensor<8xf32>
    %1 = "onnx.Conv"(%arg0, %arg1, %0) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : si64, kernel_shape = [5, 5], strides = [1, 1]} : (tensor<1x1x28x28xf32>, tensor<8x1x5x5xf32>, tensor<8xf32>) -> tensor<1x8x28x28xf32>
    return %1 : tensor<1x8x28x28xf32>
  }
}
```
Same as `onnx-mlir-opt --canonicalize --shape-inference --constprop-onnx` with implicit pass ordering removed.

**Review by AlexandreEichenberger — COMMENTED** [review]
@chentong319 high level questions: should we switch in the code places where we have `create` to `createOrFold`? I am thinking for example the the dialect builder, where this would catch most cases.

As is, the pattern that handled the folding does not seems to have been removed, as you advertised in the conversation header. Is it because there are still `create`? If so, does it mean that unless we replace all `create` to `createOrFold` we should hold on to the folding patterns?

Thanks for clarifying this for others to better understand what will be needed for the other ops.

**Comment by chentong319** [comment]
> @chentong319 high level questions: should we switch in the code places where we have `create` to `createOrFold`? I am thinking for example the the dialect builder, where this would catch most cases.
> 

I  think so. The most needed case is in onnx to krnl lowering.

> As is, the pattern that handled the folding does not seems to have been removed, as you advertised in the conversation header. Is it because there are still `create`? If so, does it mean that unless we replace all `create` to `createOrFold` we should hold on to the folding patterns?

Keep the Constant Propagation for a while. At least it contains the tree restructuring for commutative operations.
> 
> Thanks for clarifying this for others to better understand what will be needed for the other ops.

**Comment by chentong319** [comment]
@sorenlassen Sorry I forgot that you already have the draft PR #1902 for folding. My PR is more of the experiment with SqueezeOp. I will close this  PR.  However, there are several changes you may be useful for your PR.

1. Add Shape Inference  in fold. 
2. Update the const propagation for SizeOp.
3. Some lit test cases are not correct.
4. I did not add any change to the disposableDenseElementAttr.

**Comment by sorenlassen** [comment]
@chentong319 sorry for the confusion, I'm very supportive of your work in this PR and encourage you to continue, I wasn't actively working on constant folding and just closed PR #1902 to make that clear

what I meant to say in today's meeting is that I've been thinking about the underlying issues and I'd like to help, if you share details about the obstacles you encounter

**Comment by chentong319** [comment]
The current issue I have is runtime error in python for test_if_cpu in check-onnx-backend. Other questions I have are:
1. Does createOrFold work lowering onnx to krnl when unrealized cast or bufferization.to_mem/to_tensor is added? If not,  what's the solution?
2. Do we need to.keep all the different attribute for ONNXConstantOP, or just use sparseValue and Value?
3. Any change needed in the disposableDenseElementsAttr for fold as your PR name indicated?

@sorenlassen Can we schedule an online meeting to discuss how to combine the two PRs if you like to? I am available before 4 pm today.

**Comment by sorenlassen** [comment]
> Can we schedule an online meeting to discuss how to combine the two PRs

yes, that sounds good

sorry I didn't respond to your message earlier - which timezone are you in? I'm available today from now until late (5:30pm-9pm PST in California) or tomorrow after 4pm PST

**Comment by sorenlassen** [comment]
> 2\. Do we need to.keep all the different attribute for ONNXConstantOP, or just use sparseValue and Value?

I think it will be ok to just use sparseValue and Value

I think we assume in many places that only Value is used

it might make sense in the future to support sparseValue

the others can hopefully be canonicalized to Value, if we don't already

**Comment by sorenlassen** [comment]
> 3\. Any change needed in the disposableDenseElementsAttr for fold as your PR name indicated?

No.

The reason I wrote PR #1902 was because @tungld asked in the Nov 29 meeting how DisposableElementsAttr would interact with constant folding: https://github.com/onnx/onnx-mlir/wiki/Informal-meeting-agenda-and-notes#nov-29th-2022
The PR was meant to demonstrate that DisposableElementsAttr will work with constant folding (except I ran into the same issues as you with the test_if_cpu backend test and ONNXSizeOp etc).

(See: https://github.com/onnx/onnx-mlir/pull/1874#issuecomment-1339633488)

**Comment by chentong319** [comment]
> > Can we schedule an online meeting to discuss how to combine the two PRs
> 
> yes, that sounds good
> 
> sorry I didn't respond to your message earlier - which timezone are you in? I'm available today from now until late (5:30pm-9pm PST in California) or tomorrow after 4pm PST

I am on east coast. How about 5:30 pm PST today(Thursday) https://ibm.webex.com/meet/chentong90?

**Comment by sorenlassen** [comment]
> How about 5:30 pm PST today(Thursday)

that works, I'll see you then

File: src/Dialect/ONNX/ONNXOps/Tensor/Constant.cpp [hunk 5]
```diff
@@ -70,3 +70,27 @@ LogicalResult ONNXConstantOp::inferShapes(
...
+    return getValueIntAttr();
+  else if (getValueIntsAttr())
+    return getValueIntsAttr();
+  else if (getValueFloatsAttr())
+    return getValueFloatsAttr();
+  else {
+    assert(getValueStringAttr() &&
+           "ONNXConstantOp does not have a valid attribute");
+    return getValueStringAttr();
+  }
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps/Tensor/Constant.cpp:95]: we should also check getValueStringsAttr(), right?

sorry for trying to rewrite the whole method but I'd prefer if we write it as:
```
if (auto sparseValue = getSparseValueAttr()) return sparseValue;
if (auto value = getSparseValueAttr()) return value;
...
if (auto valueStrings = getValueStringsAttr()) return valueStrings;
llvm_unreachable("ONNXConstantOp does not have a valid attribute");
```
so it's consistent and symmetrical for ease of reading, and we don't look up any attribute twice

File: src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp [hunk 6]
```diff
@@ -183,3 +183,46 @@ namespace onnx_mlir {
...
+//===----------------------------------------------------------------------===//
+// Folder
+//===----------------------------------------------------------------------===//
+OpFoldResult ONNXSqueezeOp::fold(FoldAdaptor adaptor) {
+  // Fold type
+  if (failed(inferShapes(nullptr)))
+    return nullptr;
+
+  // Fold value
+
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp:196]: please remove blank line

File: src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp [hunk 7]
```diff
@@ -183,3 +183,46 @@ namespace onnx_mlir {
...
+  assert(hasStaticShape(getSqueezed().getType()) &&
+         "Shape should be static when the inputs are constant");
+
+  OnnxElementsAttrBuilder elementsBuilder(getContext());
+  return elementsBuilder.reshape(
+      adaptor.getData(), getShape(getSqueezed().getType()));
+}
+
+OpFoldResult ONNXSqueezeV11Op::fold(FoldAdaptor adaptor) {
+
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp:211]: please remove blank line

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 8]
```diff
@@ -96,78 +96,50 @@ func.func @cast_elimination(%arg0: tensor<2xf32>) -> tensor<2xf32> {
 
 // -----
 
-func.func @test_conv_batchnormtestmode_fusion_nobias(%arg0 : tensor<1x3x224x224xf32>) -> tensor<1x64x112x112xf32> {
+func.func @test_conv_batchnormtestmode_fusion_nobias(%arg0: tensor<1x3x224x224xf32>, %0: tensor<64x3x7x7xf32>, %2: tensor<64xf32>, %3: tensor<64xf32>, %4: tensor<64xf32>, %5: tensor<64xf32>) -> tensor<1x64x112x112xf32> {
```
Comment (sorenlassen) [inline test/mlir/onnx/onnx_canonicalization.mlir:99]: any reason to break with the convention to name function args `%arg1` etc?

same in line 123 below

File: test/mlir/onnx/onnx_fold.mlir [hunk 9]
```diff
@@ -0,0 +1,32 @@
+// RUN: onnx-mlir-opt --shape-inference --canonicalize %s -split-input-file | FileCheck %s
```
Comment (sorenlassen) [inline test/mlir/onnx/onnx_fold.mlir:1]: what happens if you omit --shape-inference?

File: src/Accelerators/NNPA/Dialect/ZHigh/ZHigh.td [hunk 10]
```diff
@@ -28,6 +28,7 @@ def ZHigh_Dialect : Dialect {
   let summary = "A high-level dialect for the ONNX NNPA accelerator ISA.";
   let cppNamespace = "::onnx_mlir::zhigh";
   let useDefaultAttributePrinterParser = 1;  
+  let useFoldAPI = kEmitFoldAdaptorFolder;
```
Comment (sorenlassen) [inline src/Accelerators/NNPA/Dialect/ZHigh/ZHigh.td:31]: this change and the corresponding change to ZLow looks unrelated, should they be omitted from this PR?

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 11]
```diff
@@ -211,6 +213,17 @@ ParseResult ONNXConstantOp::parse(OpAsmParser &parser, OperationState &result) {
...
+Operation *ONNXDialect::materializeConstant(
+    OpBuilder &builder, Attribute value, Type type, Location loc) {
+
+  // The atrribute could be DenseElemnemntsAttr, IntAttr, FloatAttr and etc.
+  // onnx builder is used to convert it into value()
+  OnnxBuilder onnx(builder, loc);
+  Value result = onnx.constant(value);
+  return result.getDefiningOp();
+}
+
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps.cpp:226]: please move to the end of the file so that all the print and parse methods are next to each other

and please follow the file convention and add`//===-----...` etc to the comment above the method

**Review by sorenlassen — APPROVED** [review]
Nice!

Just some stylistic comments

**Comment by tungld** [comment]
> Update the lit test. A new test case is added. For SqueezeOp, the canonicalization can do what constprop does previously

Should we remove the rule for SqueezeOp in `constprop` since its `fold` does the same thing?

**Comment by chentong319** [comment]
> > Update the lit test. A new test case is added. For SqueezeOp, the canonicalization can do what constprop does previously
> 
> Should we remove the rule for SqueezeOp in `constprop` since its `fold` does the same thing?

We will move most of the constant propagation into fold in future.
